src/tfr_control/src/arm_action_server.cpp

```diff
@@ -0,0 +1,103 @@
+ *              devel/include/tfr_msgs/ExampleGoal.h
+ *              devel/include/tfr_msgs/ExampleResult.h
+ * 
+ ***************************************************************************************/
+
+#include <actionlib/server/simple_action_server.h>
+#include <tfr_msgs/ArmMoveAction.h>
+#include <moveit/move_group_interface/move_group_interface.h>
+
+typedef actionlib::SimpleActionServer<tfr_msgs::ArmMoveAction> Server;
+
+moveit::planning_interface::MoveGroupInterface *move_group;
+const robot_state::JointModelGroup *joint_model_group;
+
+// This is the method that will be called when a client makes use
+// of this server. The provided goal is the "input"
+void execute(const tfr_msgs::ArmMoveGoalConstPtr& goal, Server* server)
+{
+    // Set up the joint space goal vector to travel to based on the input goal
+    // from the action server
+    std::vector<double> joint_group_positions(4);
+    joint_group_positions[0] = goal->pose[0];
+    joint_group_positions[1] = goal->pose[1];
+    joint_group_positions[2] = goal->pose[2];
+    joint_group_positions[3] = goal->pose[3];
+
+    // Set the current target
+    move_group->setJointValueTarget(joint_group_positions);
+    moveit::planning_interface::MoveGroupInterface::Plan my_plan;
+
+    // Try to plan to the given target
+    bool success = (move_group->plan(my_plan) == moveit::planning_interface::MoveItErrorCode::SUCCESS);
+
+    if (success) {
+        // Planning was successful, actually execute the movement
+        ROS_DEBUG("Executing movement");
+        move_group->move();
+        // TODO: How are errors handled?
+        //  - Further note: in the case that, say, the arm gets stuck and can't
+        //    move, how are these errors returned? Exceptions, return code, or
+        //    something else?
+    } else {
+        ROS_WARN("Planning of movement failed, not executing");
+    }
+
+    // Get the current position to return in the result message
+    std::vector<double> current_pos;
+    move_group->getCurrentState()->copyJointGroupPositions(joint_model_group, current_pos);
+
+    // Populate the result message
+    tfr_msgs::ArmMoveResult result;
+    result.pose.resize(4);
+    result.pose[0] = current_pos[0];
+    result.pose[1] = current_pos[1];
+    result.pose[2] = current_pos[2];
+    result.pose[3] = current_pos[3];
+
+    // Send the result message and set the appropriate action server status
+    if (success) {
+        server->setSucceeded(result);
+    } else {
+        server->setAborted(result);
+    }
+}
+
+int main(int argc, char** argv)
+{
+    ros::init(argc, argv, "arm_action_server");
+    ros::NodeHandle n;
+
+    // An async spinner is required here for the MoveIt setup to connect properly
+    ros::AsyncSpinner spinner(1);
+    spinner.start();
+
+    // Initialize the move_group connection (tl;dr sets up a ton of action server
+    // connections internally)
+    moveit::planning_interface::MoveGroupInterface mg("arm_end");
+    move_group = &mg;
```

Comment:
So you store a raw pointer as a global variable, that is used in an action server with an execute callback. 

Execute callbacks get started threaded by default so you are entering a potential shared memory "danger zone", you might not have made any mistakes, but the people maintaining your code might not be as dilligent. Global variables and threading don't play nicely. 

Probably what would make sense is to have a helper object or class that can hold
```
moveit::planning_interface::MoveGroupInterface *move_group;
const robot_state::JointModelGroup *joint_model_group;
```
as `private` member variables.


Here is how I would write it, but you do you. My one requirement is avoid the unscoped publicly accessible global variables in a threaded environment.

```
class MyClass
{
public:
  MyClass(ros::NodeHandle &n) :
      move_group{"arm_end"},
      joint_model_group{move_group.getCurrentState()->getJointModelGroup("arm_end")},
      server{n, "move_arm", boost::bind(&MyClass::execute, this, _1), false}
  { server.start(); }
  
...
private:
...
void execute(const tfr_msgs::ArmMoveGoalConstPtr& goal)
{
...
server.setSucceeded(foo);
}
...
  moveit::planning_interface::MoveGroupInterface &move_group;
  const robot_state::JointModelGroup &joint_model_group;
  actionlib::SimpleActionServer<tfr_msgs::ArmMoveAction>  &server;
}
}

int main(...)
{
...
ros::NodeHandle n;
ros::AsyncSpinner spinner(1);
spinner.start();
...
MyObj obj(n);
while (ros::ok()) ;
return 0;
}
```

Reply:
I don't think it has to be a class, but I agree that it shouldn't be stored globally here with no shared-data protections.  I don't think there's a race condition in the code you've written, but that doesn't mean it couldn't be introduced later by accident.

Reply:
What alternatives to a class would you suggest?

Reply:
I think I'm going to go with a class, @collinsc showed me that pattern earlier and it does seem really clean and helps solve some other issues I have sometimes.

Reply:
@collinsc You can also create static variables inside of free functions, which are basically scoped globals, but allow for access guards.  I don't really think that's better, but it does work.

Reply:
For the sake of my own curiosity, could you put that in pseudocode?